### PR TITLE
[NEW] Handle errors & redirect into API requests

### LINKED ISSUE #937: [NEW] Handle HTTP redirects (301, 308) and update the server list accordingly
# Description

When server changes it's URL the app needs to ask the user if user wants to change it or not, showing the new URL.

So, the app needs to do the following things:

1. Before connecting to the web-socket, check `/api/info` always and see server is valid;
2. If response is `301` or `308` status code, app needs to check if there's a `Location` header in the 
3. If `Location` is a valid URL, needs to check if it's a valid Rocket.Chat server, by requesting `/api/info`. This process is already being done on the `ConnectViewController.swift` file;
4. If new server is valid, user needs to be prompted with an alert asking if user want's to change the old URL to the new one;
5. If user taps "Yes", the app needs to clear all the database, change server information to the new URL and keep the auth information;

### PR DESCRIPTION
**Pull request by rafaelks**
@RocketChat/ios

Closes #937

## Description

This PR is suppose to add a way to handle errors and redirect to all of our REST API calls. It will also add an alert when we request `/api/info` and server tries to redirect, alerting user that server's URL changed.

## Progress

- [x] Add error & statusCode to APIResult
- [x] Call completion when URLSession notifies a redirect
- [x] i18n
- [x] Unit tests

### CODE REVIEW

File: Rocket.Chat/Controllers/Chat/MembersListViewController.swift [hunk 1]
```diff
@@ -38,7 +38,7 @@ class MembersListViewData {
 
         if let subscription = subscription {
             isLoadingMoreMembers = true
-            API.shared.fetch(SubscriptionMembersRequest(roomId: subscription.rid, type: subscription.type), options: .paginated(count: pageSize, offset: currentPage*pageSize)) { result in
+            API.shared.fetch(SubscriptionMembersRequest(roomId: subscription.rid, type: subscription.type), options: .paginated(count: pageSize, offset: currentPage*pageSize), { result in
```
Comment (cardoso) [inline Rocket.Chat/Controllers/Chat/MembersListViewController.swift]: Can we do the same thing here :top:

File: Rocket.Chat/Controllers/Chat/MessagesListViewController.swift [hunk 2]
```diff
@@ -53,8 +53,11 @@ class MessagesListViewData {
 
         if let subscription = subscription {
             isLoadingMoreMessages = true
-            API.shared.fetch(SubscriptionMessagesRequest(roomId: subscription.rid, type: subscription.type, query: query),
-                             options: .paginated(count: pageSize, offset: currentPage*pageSize)) { result in
+
+            let request = SubscriptionMessagesRequest(roomId: subscription.rid, type: subscription.type, query: query)
+            let options = APIRequestOptions.paginated(count: pageSize, offset: currentPage*pageSize)
+
+            API.shared.fetch(request, options: options, { result in
```
Comment (cardoso) [inline Rocket.Chat/Controllers/Chat/MessagesListViewController.swift]: That you did here 🔝 ?

File: Rocket.Chat/Managers/Requests/InfoRequestHandler.swift [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        if let location = response.allHeaderFields["Location"] as? String {
+            if let newURL = transformNewURL(location) {
+                handleRedirect(newURL)
+            }
+        }
+
+        completionHandler(nil)
+    }
+
+    func transformNewURL(_ newURL: String) -> URL? {
```
Comment (cardoso) [inline Rocket.Chat/Managers/Requests/InfoRequestHandler.swift]: We have URL.init(string:, scheme:) for this. Can we remove this and use that instead?

File: Rocket.ChatTests/Managers/DatabaseManagerSpec.swift [hunk 4]
```diff
@@ -9,26 +9,40 @@
 import XCTest
 @testable import Rocket_Chat
 
+private let testServers = [[
```
Comment (cardoso) [inline Rocket.ChatTests/Managers/DatabaseManagerSpec.swift]: Is this global 😱 ? Should we move it to the DatabaseManager extension?

Comment (rafaelks) [reply Rocket.ChatTests/Managers/DatabaseManagerSpec.swift]: You actually can't add properties to the extensions... let's leave it there until we find a better solution?

File: Rocket.Chat/Controllers/MainViewController.swift [hunk 5]
```diff
@@ -98,4 +76,82 @@ final class MainViewController: BaseViewController {
...
+        DatabaseManager.changeDatabaseInstance()
+        AuthManager.recoverAuthIfNeeded()
+
+        DispatchQueue.main.async {
+            if let auth = AuthManager.isAuthenticated() {
+                if let apiHost = auth.apiHost {
+                    self.infoRequestHandler.validate(with: apiHost)
+                }
+            }
+        }
```
Comment (cardoso) [inline Rocket.Chat/Controllers/MainViewController.swift]:
```swift
DispatchQueue.main.async {
    guard let auth = AuthManager.isAuthenticated() else { return }
    guard let apiHost = auth.apiHost() else { return }
    self.infoRequestHandler.validate(with: apiHost)
 }
```

File: Rocket.Chat/Controllers/Chat/MembersListViewController.swift (hunk 1 above)
Comment (rafaelks) [reply Rocket.Chat/Controllers/Chat/MembersListViewController.swift]: Sure thing!

File: Rocket.ChatTests/Managers/DatabaseManagerSpec.swift (hunk 4 above)
Comment (cardoso) [reply Rocket.ChatTests/Managers/DatabaseManagerSpec.swift]: Always forget that 👍

File: Rocket.Chat/Managers/Requests/InfoRequestHandler.swift (hunk 3 above)
Comment (rafaelks) [reply Rocket.Chat/Managers/Requests/InfoRequestHandler.swift]: Done, good finding!

File: Rocket.Chat/Controllers/MainViewController.swift (hunk 5 above)
Comment (rafaelks) [reply Rocket.Chat/Controllers/MainViewController.swift]: Done!

File: Rocket.Chat/Controllers/Chat/MessagesListViewController.swift (hunk 2 above)
Comment (rafaelks) [reply Rocket.Chat/Controllers/Chat/MessagesListViewController.swift]: Done! 👍

**Pull request merged**